src/Microsoft.Azure.ServiceBus/MessageBatchReceivePump.cs

```diff
@@ -0,0 +1,277 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.Azure.ServiceBus
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Core;
+    using Primitives;
+
+    sealed class MessageBatchReceivePump
+    {
```

Comment:
There's a LOT of duplication here. Since this is an internal class, we don't need it to be explicit about batching. I'd recommend doing the changes directly in the base class by handling all the cases.